### PR TITLE
Fix issue with visibility of modification in DNSCache and docs cleanup in SimpleLockManager

### LINKED ISSUE #186: OOM and high CPU caused by door camera
Hi all

I have a strange problem caused by our door bell camera discovered through mDNS
Problem eventually leads to 100% CPU and OutOfMemory errors on my JVM.

Camera sends packet which looks like this

`
000084000000000e00000000184772616e6473747265616d2d303030423832423834364434056c6f63616c00000180010000007800040b0100320235300130013102313107696e2d6164647204617261700000018001000000780002c00c224772616e6473747265616d2047445333373130202d20303030423832423834364434055f68747470045f746370056c6f63616c00002180010000007800080000000001bbc00cc05e0010800100001194000100095f7365727669636573075f646e732d7364045f756470056c6f63616c00000c0001000011940002c081c081000c0001000011940002c05e224772616e6473747265616d2047445333373130202d20303030423832423834364434055f72747370045f746370056c6f63616c000021800100000078000800000000022ac00cc0ea0010800100001194000100c0b2000c0001000011940002c0eac0ea000c0001000011940002c0ea224772616e6473747265616d2047445333373130202d20303030423832423834364434125f4772616e6473747265616d2d766964656f045f746370056c6f63616c00002180010000007800080000000001bbc00cc15a0010800100001194000100c0b2000c0001000011940002c0eac0ea000c0001000011940002c15a
`

This can be broken down as follows (bulk at the end omitted)

```
Some header
000084000000000e00000000 

First Type A answer
18 4772616e6473747265616d2d303030423832423834364434056c6f63616c00000180010000007800040b010032

2nd Type A record
02 35300130013102313107696e2d61646472046172617000000180010000007800 02 c00c
```
I have no problem with first answer
`[IPv4Address@1341712829 type: TYPE_A index 1, class: CLASS_IN index 1-unique, name: Grandstream-000B82B846D4.local. ttl: '0/120' address: '11.1.0.50']`

But the 2nd one is funny. Last part 02 c00c indicates there are 2 bytes for records data, and that really should be 4 bytes as far as I can tell as this is Inet4Address

This looks like a bug in the camera. We've upgraded to the latest firmware but the problem stays. We will be in touch with manufacturer as well.

The thing is JmDNS still creates record but with IP which is null.
```
java.net.UnknownHostException: addr is of illegal length
	at java.net.InetAddress.getByAddress(InetAddress.java:1042)
	at java.net.InetAddress.getByAddress(InetAddress.java:1439)
	at javax.jmdns.impl.DNSRecord$Address.<init>(DNSRecord.java:307)
	at javax.jmdns.impl.DNSRecord$IPv4Address.<init>(DNSRecord.java:211)
	at javax.jmdns.impl.DNSIncoming.readAnswer(DNSIncoming.java:326)
	at javax.jmdns.impl.DNSIncoming.<init>(DNSIncoming.java:234)
	at javax.jmdns.impl.SocketListener.run(SocketListener.java:50)
```

Later on more records are added they expire because they never match through equality check

```
INFO  2019-04-03 15:18:53.796  SocketListener(DESKTOP-D5IPNMD.local.)   Inv   sameValue()                    Failed to compare addresses of DNSRecords      (DNSRecord.java:336)
java.lang.NullPointerException: null
	at javax.jmdns.impl.DNSRecord$Address.sameValue(DNSRecord.java:334)
	at javax.jmdns.impl.JmDNSImpl.handleRecord(JmDNSImpl.java:1384)
	at javax.jmdns.impl.JmDNSImpl.handleResponse(JmDNSImpl.java:1462)
	at javax.jmdns.impl.SocketListener.run(SocketListener.java:61)
```

Also eviction does not happen 
```
INFO  2019-04-03 15:20:00.626  JmDNS(DESKTOP-D5IPNMD.local.).Timer      Inv   sameValue()                    Failed to compare addresses of DNSRecords      (DNSRecord.java:336)
java.lang.NullPointerException: null
	at javax.jmdns.impl.DNSRecord$Address.sameValue(DNSRecord.java:334)
	at javax.jmdns.impl.DNSRecord.equals(DNSRecord.java:66)
	at java.util.ArrayList.remove(ArrayList.java:510)
	at javax.jmdns.impl.DNSCache.removeDNSEntry(DNSCache.java:237)
	at javax.jmdns.impl.JmDNSImpl.cleanCache(JmDNSImpl.java:1888)
	at javax.jmdns.impl.tasks.RecordReaper.run(RecordReaper.java:56)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
```

This causes pile up of entries and soon very high CPU due to entry comparison, exceptions thrown and logged. In my case SLF4J was configured to NOPLogger for some reason, which may contributed to 100% CPU due to no-ops.

My understanding of DNS is quite vague, but it does look like a firmware bug.
- is it possible that there is a flavour of Type A answer which can use 2 bytes only?
- is there anything we could do to fix this and avoid ripple effect in JmDNS library?

My thought would be to do something like that in DNSIncoming line 326:

```
                if (len == 4) 
                    rec = new DNSRecord.IPv4Address(domain, recordClass, unique, ttl, _messageInputStream.readBytes(len));
                else 
                    _messageInputStream.skip(len);
```

I appreciate this is solving problem created by something else (if it is in fact firmware problem), but I would like to avoid mentioned problems.

Regards
Daniel

### PR DESCRIPTION
**Pull request by ivenhov**
Fix #186 

This is to based on discussions in the comments onwards
https://github.com/jmdns/jmdns/issues/186#issuecomment-2308543051

Related to potential lack of visibility of mutations in the entryLists between threads

### CODE REVIEW

File: src/main/java/javax/jmdns/impl/DNSCache.java [hunk 1]
```diff
@@ -52,7 +52,7 @@ public class DNSCache extends ConcurrentHashMap<String, List<DNSEntry>> {
 
     private static final long   serialVersionUID    = 3024739453186759259L;
     
-    private final SimpleLockManager _lm = new SimpleLockManager();
+    private final transient SimpleLockManager _lm = new SimpleLockManager();
```
Comment (joerg1985) [inline src/main/java/javax/jmdns/impl/DNSCache.java:55]: `transistent` is a keyword to change behavoiur of serialization. This field is final so this is thread safe without any further keyword.

Comment (ivenhov) [reply src/main/java/javax/jmdns/impl/DNSCache.java:55]: Final fields are also serialized by default, unless static as far as I know.

Comment (joerg1985) [reply src/main/java/javax/jmdns/impl/DNSCache.java:55]: I did assume this change is for thread safty?

Comment (ivenhov) [reply src/main/java/javax/jmdns/impl/DNSCache.java:55]: Pull request yes. However I crammed in transient at the same time when I realised DNSCache is Serializable.

File: src/main/java/javax/jmdns/impl/DNSCache.java [hunk 2]
```diff
@@ -218,10 +218,10 @@ public boolean addDNSEntry(DNSEntry dnsEntry) {
             if (entryList == null) {
                 entryList = new ArrayList<>(3);
                 entryList.add(dnsEntry);
```
Comment (kaikreuzer) [inline src/main/java/javax/jmdns/impl/DNSCache.java:220]: I assume you also wanted to delete this line?

**Review by kaikreuzer — CHANGES_REQUESTED** [review]

File: src/main/java/javax/jmdns/impl/DNSCache.java (hunk 2 above)
Comment (ivenhov) [reply src/main/java/javax/jmdns/impl/DNSCache.java:220]: Sorry Kai, which line you are referring to?

Comment (kaikreuzer) [reply src/main/java/javax/jmdns/impl/DNSCache.java:220]: The one where I placed the comment: line 220. It looks redundant, since you've added line 222.

Comment (ivenhov) [reply src/main/java/javax/jmdns/impl/DNSCache.java:220]: Yes, please remove. Good catch, thanks!

**Comment by melloware** [comment]
@ivenhov if you change your first comment text in this thread to `Fix #186` it will link this PR to the ticket

File: src/main/java/javax/jmdns/impl/DNSCache.java (hunk 2 above)
Comment (kaikreuzer) [inline src/main/java/javax/jmdns/impl/DNSCache.java:220]:
```suggestion
```

Comment (kaikreuzer) [reply src/main/java/javax/jmdns/impl/DNSCache.java:220]: > Yes, corrected.

Where exactly? I still see it...

Comment (ivenhov) [reply src/main/java/javax/jmdns/impl/DNSCache.java]: Number of commits in the pull request changed from 3 back to 2 again. I've accepted yours

**Comment by melloware** [comment]
@ivenhov looks like you need to re-sign the commit

**Review by kaikreuzer — APPROVED** [review]
Thanks!

**Comment by kaikreuzer** [comment]
> looks like you need to re-sign the commit

No need, all fine - the DCO check was only red because my change was added and GitHub uses a "co-authored" line on such commits.

**Pull request merged**